spec/models/auth_policies/project_policy_spec.rb

```diff
@@ -0,0 +1,242 @@
+    }
+  end
+
+  let(:role) { create(:role, **permissions.compact_blank) }
+
+  let(:hud_data_access_role) { create(:role, can_upload_hud_zips: true, can_edit_data_sources: true) }
+
+  shared_examples 'permission checks' do |role_present|
+    it 'handles basic permissions appropriately' do
+      [
+        [:can_edit?, false],
+        [:can_delete?, false],
+        [:can_view?, true],
+        [:can_view_imports?, false],
+        [:can_view_clients?, true],
+        [:can_view_project_locations?, false],
+      ].each do |method, expected|
+        expected = role_present ? expected : false
+        actual = policy.send(method)
+        expect(actual).to eq(expected), "#{method}: #{actual} != #{expected}"
+      end
+    end
+
+    context 'with no data source permissions' do
+      it 'denies access to raw HMIS data' do
+        expect(policy.can_see_raw_hmis_data?).to be false
+      end
+    end
+
+    context 'project name visibility' do
+      context 'with non-confidential project' do
+        it 'allows viewing project name' do
+          expected = role_present
+          expect(policy.can_view_name?).to eq(expected)
+        end
+      end
+
+      context 'with confidential project' do
+        it 'respects confidential project name permissions' do
+          project.update!(confidential: true)
+          expected = role_present && permissions[:can_view_confidential_project_names]
+          expect(policy.can_view_name?).to eq(expected)
+        end
+      end
+    end
+  end
+
+  context 'with legacy user permissions' do
+    let(:access_group) { create(:access_group) }
+    let(:user) do
+      user = create(:user)
+      role.add(user)
+      access_group.add(user)
+      user
+    end
+    let(:policy) { user.policy_for(project) }
+
+    context 'with full data source permissions' do
+      before do
+        access_group.add_viewable(data_source)
+        hud_data_access_role.add(user)
+      end
+      it 'allows access to raw HMIS data' do
+        expect(policy.can_see_raw_hmis_data?).to be true
+      end
+    end
+
+    context 'with direct project access' do
+      before { access_group.add_viewable(project) }
+      include_examples 'permission checks', true
+    end
+
+    context 'with organization access' do
+      before { access_group.add_viewable(organization) }
+      include_examples 'permission checks', true
+    end
+
+    context 'with data source access' do
+      before { access_group.add_viewable(data_source) }
+      include_examples 'permission checks', true
+    end
+
+    context 'with project group access' do
+      let(:project_group) do
+        group = create(:project_access_group)
+        project.project_groups << group
+        project.save!
+        group
+      end
+
+      before { access_group.add_viewable(project_group) }
+      include_examples 'permission checks', true
+    end
+
+    context 'with CoC code access' do
+      let(:coc_code) { 'authtest1' }
```

Comment:
It's probably fine to include a non-valid CoC Code, but at various points we've enforced CoC-Code validity, which might cause this test to fail unnecessarily in the future.  To be sure it won't get cleaned up through CoC code validation, you can use `XX-500` or `XX-501`.